torchtune/models/phi4/_tokenizer.py

```diff
@@ -0,0 +1,293 @@
+        continue
+    PHI4_SPECIAL_TOKENS[f"<|dummy_{87 - (100350 - token_id)}|>"] = token_id + 1
+
+CL100K_PATTERN = r"""(?i:'s|'t|'re|'ve|'m|'ll|'d)|[^\r\n\p{L}\p{N}]?\p{L}+|\p{N}{1,3}| ?[^\s\p{L}\p{N}]+[\r\n]*|\s*[\r\n]+|\s+(?!\S)|\s+"""  # noqa
+
+
+class Phi4MiniTokenizer(ModelTokenizer, Transform):
+    """
+    TikToken tokenizer configured with Phi4 (14B) special tokens.
+
+    Args:
+        merges_path (str): Path to merges.txt file.
+        vocab_path (str): Path to vocab.json file.
+        special_tokens (Optional[Dict[str, int]]): mapping containing special text tokens and
+            their registered token IDs. If left as None, this will be set to the canonical
+            Phi4 special tokens.
+        max_seq_len (Optional[int]): A max sequence length to truncate tokens to.
+            Default: None
+        prompt_template (Optional[PromptTemplate]): template used to format the messages based on their role. This is used
+            to add structured text around the actual messages. The structured text is used in three scenarios:
+
+            - Task-specific templates to gear models for a particular task that it will expect after training
+            - Model-specific templates that are required whenever the model is prompted, such as the [INST]
+              tags in Llama2 and in Mistral
+            - Community standardized templates, such as :class:`~torchtune.data.ChatMLTemplate`
+
+            The extra text will still get tokenized as normal text, not as special tokens. Default is None.
+
+    Examples:
+        >>> tokenizer = Phi4MiniTokenizer(vocab_path="vocab.json", merges_path="merges.txt")
+        >>> tokenized_text = tokenizer.encode("Hello world!", add_bos=True, add_eos=True)
+        >>> print(tokenized_text)
+        [1, 31587, 29644, 102, 2]
+    """
+
+    def __init__(
+        self,
+        merges_path: str = None,
+        vocab_path: str = None,
+        special_tokens: Optional[Dict[str, int]] = None,
+        max_seq_len: Optional[int] = None,
+        prompt_template: Optional[PromptTemplate] = None,
+    ):
+        self.special_tokens = (
+            special_tokens if special_tokens is not None else PHI4_SPECIAL_TOKENS
+        )
+
+        # Use custom EOS, BOS and pad ids instead of TikToken's
+        self.eos_id = self.special_tokens["<|im_end|>"]
+        self.bos_id = self.special_tokens["<|endoftext|>"]
+        self.pad_id = self.special_tokens["<|dummy_85|>"]
+
+        # During generation, stop when eos_id is encountered
+        self.stop_tokens = [self.eos_id]
+
+        self.max_seq_len = max_seq_len
+
+        self.prompt_template = prompt_template
+
+        self.tokenizer_model = GPT2BaseTokenizer(
+            vocab_path,
+            merges_path,
+            "replace",
+            self.eos_id,
+            self.bos_id,
+            self.eos_id,
+            self.pad_id,
+        )
+
+    @property
+    def vocab_size(self):
+        return self.tt_model.vocab_size
```

Comment:
I think this needs to be updated
```suggestion
        return self.tokenizer_model.vocab_size
```